provisioning/baremetal_config.go

```diff
@@ -140,36 +139,53 @@ func getProvisioningIPCIDR(config *metal3iov1alpha1.ProvisioningSpec) *string {
 	return nil
 }
 
+// joinHostIPPort infers the IP version from the machine network CIDR, and wraps it as appropriate. When
+// the provisioning network is disabled, we dynamically set PROVISIONING_IP to the Kubernetes node hostIP using
+// a field selector. However, we have to let Kubernetes do the variable interpolation for us, but its not smart
+// enough to do conditional wrapping of IPv6 URL's, thus we infer it from the machine CIDR.
+func joinHostIPPort(config *metal3iov1alpha1.ProvisioningSpec, port string) string {
+	_, net, err := net.ParseCIDR(config.ProvisioningNetworkCIDR)
+	if err == nil {
+		if net.IP.To4() == nil {
+			return fmt.Sprintf("[$(PROVISIONING_IP)]:%s", port)
+		}
+	}
+
+	return fmt.Sprintf("$(PROVISIONING_IP):%s", port)
+}
```

Comment:
My whole solution for this is a disaster.

Could we query the Kubernetes API somewhere else to get the node's host IP and pass that through to here? I didn't find a logical place to do it.

Still investigating but open to ideas.

Reply:
If we were to run [ironic-ipa-downloader](https://github.com/openshift/ironic-ipa-downloader/blob/master/get-resource.sh) (with some modifications to ensure it doesn't download multiple times from the Internet) in the [second-level image cache](https://github.com/zaneb/cluster-baremetal-operator/commits/image-cache) then we could refer to this file in the second-level cache using the API VIP when there is no provisioning VIP. This is probably a good thing anyway as it would mean we never have to re-download IPA from the Internet once we have it.

Out of curiousity, can anybody explain why we need to both [set this path](https://github.com/metal3-io/ironic-image/blob/master/runhttpd.sh#L34-L37) in [inspector.pxe](https://github.com/metal3-io/ironic-image/blob/master/inspector.ipxe) *and* [pass it to ironic](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L294-L297) from the BMO?

Reply:
> f we were to run ironic-ipa-downloader (with some modifications to ensure it doesn't download multiple times from the Internet) in the second-level image cache then we could refer to this file in the second-level cache using the API VIP when there is no provisioning VIP. 

Indeed, this is simpler if we get the second-level stuff up and running. If we wanted this solution to work by itself  though we need a way to generate the URL.

I'm pretty sure I could query for pod for the IP somewhere in the early controller start-up code? I'll play around with that today.

> Out of curiousity, can anybody explain why we need to both set this path in inspector.pxe and pass it to ironic from the BMO?

Just a guess, but I think it uses inspector.pxe for PXE booting, and Ironic itself needs it for virtuamedia boots

Reply:
> I'm pretty sure I could query for pod for the IP somewhere in the early controller start-up code? I'll play around with that today.

Angus already wrote it  :) 
https://github.com/openshift/cluster-baremetal-operator/commit/c2ab117ed3001e72389f031d7b4fde28f3c96ae1